File: debop4s-timeperiod/src/test/scala/debop4s/timeperiod/tests/timeranges/MinuteRangeCollectionTest.scala
package debop4s.timeperiod.tests.timeranges


import debop4s.timeperiod._
import debop4s.timeperiod.tests.AbstractTimePeriodTest
import debop4s.timeperiod.timerange.MinuteRangeCollection
import debop4s.timeperiod.utils.Times


/**
 * MinuteRangeCollectionTest
 * Created by debop on 2014. 2. 16.
 */
class MinuteRangeCollectionTest extends AbstractTimePeriodTest {

  test("single minute") {
    val now = Times.now
    val minutes = new MinuteRangeCollection(now, 1, EmptyOffsetTimeCalendar)

    val startTime = Times.trimToSecond(now)
    val endTime = Times.trimToSecond(now) + 1.minute

    minutes.minuteCount should equal(1)

    minutes.start should equal(startTime)
    minutes.end should equal(endTime)

    val mins = minutes.minutes
    mins.size should equal(1)
    mins(0).start should equal(startTime)
    mins(0).end should equal(endTime)
  }

  test("calendar minutes") {
    val now = Times.now

    for (m <- 1 until 100 by 5) {
      val minutes = MinuteRangeCollection(now, m)

      val startTime = Times.trimToSecond(now)
      val endTime = Times.trimToSecond(now).plusMinutes(m).plus(minutes.calendar.endOffset)

      minutes.minuteCount should equal(m)
      minutes.start should equal(startTime)
      minutes.end should equal(endTime)

      val items = minutes.minutes

      for (i <- 0 until m) {
        items(i).start should equal(startTime + i.minute)
        items(i).unmappedStart should equal(startTime + i.minute)

        items(i).end should equal(minutes.calendar.mapEnd(startTime + (i + 1).minute))
        items(i).unmappedEnd should equal(startTime + (i + 1).minute)
      }
    }
  }

  test("minutes") {
    val minuteCounts = Array(1, 24, 48, 64, 128)
    val now = Times.now

    minuteCounts.foreach {
      minuteCount =>
        val minuteRanges = MinuteRangeCollection(now, minuteCount)
        val startTime = Times.trimToSecond(now).plus(minuteRanges.calendar.startOffset)
        val endTime = startTime.plusMinutes(minuteCount).plus(minuteRanges.calendar.endOffset)

        minuteRanges.start should equal(startTime)
        minuteRanges.end should equal(endTime)
        minuteRanges.minuteCount should equal(minuteCount)

        val items = minuteRanges.minutes
        items.size should equal(minuteCount)

        (0 until minuteCount).par.foreach {
          m =>
            items(m).start should equal(startTime + m.minute)
            items(m).end should equal(minuteRanges.calendar.mapEnd(startTime + (m + 1).minute))
            items(m).unmappedEnd should equal(startTime + (m + 1).minute)
        }
    }
  }

}

File: debop4s-timeperiod/src/test/java/debop4s/timeperiod/tests/timeranges/MinuteRangeCollectionTest.java
package debop4s.timeperiod.tests.timeranges;

import debop4s.core.JAction1;
import debop4s.core.parallels.JParallels;
import debop4s.timeperiod.TimeCalendar;
import debop4s.timeperiod.tests.TimePeriodTestBase;
import debop4s.timeperiod.timerange.MinuteRange;
import debop4s.timeperiod.timerange.MinuteRangeCollection;
import debop4s.timeperiod.utils.Times;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;
import org.junit.Test;

import java.util.List;

import static org.fest.assertions.Assertions.assertThat;

@Slf4j
public class MinuteRangeCollectionTest extends TimePeriodTestBase {

    @Test
    public void singleMinutes() {
        DateTime now = Times.now();
        MinuteRangeCollection minutes = new MinuteRangeCollection(now, 1, TimeCalendar.getEmptyOffset());

        DateTime startTime = Times.trimToSecond(now);
        DateTime endTime = Times.trimToSecond(now).plusMinutes(1);

        assertThat(minutes.getMinuteCount()).isEqualTo(1);

        assertThat(minutes.getStart()).isEqualTo(startTime);
        assertThat(minutes.getEnd()).isEqualTo(endTime);

        List<MinuteRange> mins = minutes.minutes();

        assertThat(mins.size()).isEqualTo(1);
        assertThat(mins.get(0).getStart()).isEqualTo(startTime);
        assertThat(mins.get(0).getEnd()).isEqualTo(endTime);
    }

    @Test
    public void calendarMinutes() {
        DateTime now = Times.now();

        for (int m = 1; m < 97; m += 5) {
            MinuteRangeCollection minutes = new MinuteRangeCollection(now, m);

            DateTime startTime = Times.trimToSecond(now);
            DateTime endTime = Times.trimToSecond(now).plusMinutes(m).plus(minutes.getTimeCalendar().endOffset());

            assertThat(minutes.getMinuteCount()).isEqualTo(m);
            assertThat(minutes.getStart()).isEqualTo(startTime);
            assertThat(minutes.getEnd()).isEqualTo(endTime);

            List<MinuteRange> items = minutes.minutes();

            for (int i = 0; i < m; i++) {
                assertThat(items.get(i).getStart()).isEqualTo(startTime.plusMinutes(i));
                assertThat(items.get(i).getUnmappedStart()).isEqualTo(startTime.plusMinutes(i));

                assertThat(items.get(i).getEnd()).isEqualTo(minutes.getTimeCalendar().mapEnd(startTime.plusMinutes(i + 1)));
                assertThat(items.get(i).getUnmappedEnd()).isEqualTo(startTime.plusMinutes(i + 1));
            }
        }
    }

    @Test
    public void minutesTest() {

        final int[] minuteCounts = new int[] { 1, 24, 48, 64, 128 };
        final DateTime now = Times.now();

        for (int minuteCount : minuteCounts) {

            final MinuteRangeCollection minuteRanges = new MinuteRangeCollection(now, minuteCount);
            final DateTime startTime = Times.trimToSecond(now).plus(minuteRanges.getTimeCalendar().startOffset());
            final DateTime endTime = startTime.plusMinutes(minuteCount).plus(minuteRanges.getTimeCalendar().endOffset());

            assertThat(minuteRanges.getStart()).isEqualTo(startTime);
            assertThat(minuteRanges.getEnd()).isEqualTo(endTime);
            assertThat(minuteRanges.getMinuteCount()).isEqualTo(minuteCount);

            final List<MinuteRange> items = minuteRanges.minutes();
            assertThat(items.size()).isEqualTo(minuteCount);

            JParallels.run(minuteCount, new JAction1<Integer>() {
                @Override
                public void perform(Integer m) {
                    assertThat(items.get(m).getStart()).isEqualTo(startTime.plusMinutes(m));
                    assertThat(items.get(m).getEnd()).isEqualTo(minuteRanges.getTimeCalendar().mapEnd(startTime.plusMinutes(m + 1)));
                    assertThat(items.get(m).getUnmappedEnd()).isEqualTo(startTime.plusMinutes(m + 1));
                }
            });
        }
    }
}
